src/rcc.rs

```diff
@@ -286,91 +293,75 @@ impl CFGR {
-            let allowed_pllp: [u8; 4] = [2, 4, 6, 8];
-            let pllp_val = *allowed_pllp
-                .iter()
-                .min_by_key(|&pllp| {
-                    plln = ((sysclk_mhz * (*pllp as f32)) / vco_clkin_mhz).floor();
-                    let error = sysclk_mhz - ((plln / (*pllp as f32)) * vco_clkin_mhz);
-
-                    if error < 0.0 || plln * vco_clkin_mhz > 432.0 || plln > 432.0 || plln < 100.0 {
-                        core::u32::MAX
-                    } else {
-                        (error * 1_000.0) as u32
-                    }
-                })
-                .unwrap();
+            // Sysclk output divisor must be one of 2, 4, 6 or 8
+            let sysclk_div = core::cmp::min(8, (432_000_000 / sysclk) & !1);
 
-            // PLLN corresponding to the best pllp_val
-            plln = ((sysclk_mhz * (pllp_val as f32)) / vco_clkin_mhz).floor();
+            // Input divisor from PLL source clock, must result to frequency in
+            // the range from 1 to 2 MHz
+            let pllm_min = (sysclk + 1_999_999) / 2_000_000;
+            let pllm_max = sysclk / 1_000_000;
 
-            // Pllp bits to be written in the register
-            let pllp = match pllp_val {
-                2 => 0b00,
-                4 => 0b01,
-                6 => 0b10,
-                8 => 0b11,
-                _ => unreachable!(),
+            let target_freq = if self.pll48clk {
+                48_000_000
+            } else {
+                sysclk * sysclk_div
             };
 
-            // Update the real sysclk value
-            sysclk_mhz = (vco_clkin_mhz * plln) / (pllp_val as f32);
-            sysclk = (sysclk_mhz * 1_000_000.0) as u32;
-
-            // Turn PLL off
-            rcc.cr.modify(|_, w| w.pllon().off());
-            // Wait till PLL is disabled
-            while !rcc.cr.read().pllrdy().is_not_ready() {}
-
-            if self.hse.is_some() {
-                // If HSE is provided
-                // Configure PLL from HSE
-                rcc.pllcfgr.write(|w| unsafe {
-                    w.pllsrc()
-                        .hse()
-                        .pllm()
-                        .bits(pllm as u8)
-                        .plln()
-                        .bits(plln as u16)
-                        .pllp()
-                        .bits(pllp)
-                        .pllq()
-                        .bits(9)
-                });
+            // Find the lowest pllm value that minimize the difference between
+            // target frequency and the real vco_out frequency.
+            let pllm = (pllm_min..=pllm_max)
+                .min_by_key(|pllm| {
+                    let vco_in = sysclk / pllm;
+                    let plln = target_freq / vco_in;
+                    target_freq - vco_in * plln
+                })
+                .unwrap();
+
+            let vco_in = sysclk / pllm;
+            assert!(vco_in >= 1_000_000 && vco_in <= 2_000_000);
```

Comment:
According to the datasheet the valid input frequency range for the PLLs is 0.95 to 2.10 MHz.

Reply:
The software has to set these bits correctly to ensure that the VCO input frequency
ranges from 1 to 2 MHz. It is recommended to select a frequency of 2 MHz to limit
PLL jitter.